nutils/numeric.py

```diff
@@ -570,4 +624,70 @@ def levicivita(n: int, dtype=float):
   result[tuple(I)] = 1 - 2*(nperms % 2)
   return result
 
+class wrapped:
+  '''hashable array container.
+
+  The container can be used for fast equality checks and for dictionary keys.
+  Data is copied at construction and canonicalized by casting it to the
+  platform's primary data representation (e.g. int64 i/o int32). It can be
+  retrieved via the `.asarray` attribute. Additionally the wrapped object
+  provide direct access to a subset of the array's attributes.
+
+  Parameters
+  ----------
+  array : `array_like` or :class:`wrapped`
+      The array or previously wrapped data.
+
+  Example
+  -------
+  >>> a = numpy.array([1,2,3])
+  >>> w = wrapped(a)
+  >>> w == wrapped([1,2,4]) # we cannot do this on a
+  False
+  >>> w.asarray
+  array([1, 2, 3])
+  '''
+
+  __slots__ = 'dtype', 'shape', 'bytes'
+
+  def __init__(self, array):
```

Comment:
Is it better to implement a `__new__` instead so that for `b = wrapped(np.array(...))` we have `wrapped(b) is b`.

Reply:
In this case `__new__` is a bit messy in combination with pickle as that calls `__new__` without arguments followed by `__setstate__` (unless we use `__reduce__`, but then we need to make sure that we do not mistake the state tuple for array contents). Deduplication does matter, though - see the new command in `__setstate__`, I solved it now with a separate function `aswrapped` which I think is the most straightforward way out.

Reply:
The latest implementation of `numeric.wrapped` builds on `types.Immutable`.

Reply:
Update: added a commit to have `types.Immutable` support `__new__`, removing the need for `aswrapped`.